source/portfolio/portfolio_mad.h

```diff
@@ -0,0 +1,25 @@
+//
+// Created by eduardo on 15/04/2021.
+//
+
+#ifndef PORTFOLIO_PORTFOLIO_MAD_H
+#define PORTFOLIO_PORTFOLIO_MAD_H
+#include "market_data_mad.h"
+#include "portfolio.h"
+#include <ostream>
+
+namespace portfolio {
+    class portfolio_mad : public portfolio {
```

Comment:
Parece que o conceito que você está fazendo aqui é mais um `portfolio_with_mad` (um portfólio com seu mad) do que um `portfolio_mad` (o mad de um portfólio). Isso acaba respondendo até algumas questões que tive em outros comentários.

Uma relação de herança para isso, em vez de composição, ficaria muito complexa de maneira desnecessária. Por exemplo, podemos ter uma classe `lago` e outra classe `pato`. Um lago pode ter 0 ou mais patos. Normalmente representaríamos isso com um vector de `pato`s no `lago` em vez de criar uma classe `lago` e outra classe `lago_com_patos`.

Basta imaginar o que aconteceria se criássemos uma classe nova ligar ou desligar cada tipo de composição que podemos ou não termos calculado. Por exemplo, teríamos `portfolio`, `portfolio_with_risk`, `portfolio_with_return`, `portfolio_with_assets`... Para cada nova métrica que for colocar no seu código, teria uma classe nova. Como cada métrica pode, no mínimo, existir ou não existir, você teria 2^n classes para representar n propriedades que podem ou não existir. Se tiver 5 métricas no futuro, vai ter que manter e testar 2^5=32 classes diferentes.

A sua relação aqui é de composição ou de simples associação. Começando por composição, você pode sizer que o portfólio terá ou não um mad_data com os dados que precisa para calcular o portfolio_mad. Assim como o lago pode ter zero ou mais patos, seu portfolio pode ter 0 ou 1 mad_data. Assim, seria uma relação de composição. Como é apenas 0 ou 1, em vez de 0 ou mais, em vez de usar um `std::vector` você vai usar o `std::optional`.

Depois disso, o portfolio_mad pode ser apenas o mad de um portfólio e não um portfolio com o mad, para evitar o problema acima. Assim, isso seria apenas uma relação de associação. A classe portfolio teria uma função para retornar seu `mad`. Quando chamada, ela retornaria esse `mad` e pronto.

Reply:
Fiz a alteração no código para adequar a estrutura geral do portfolio, portfolio_mad, market_data e market_data_mad. Seguindo as sugestões, portfolio_mad deixou de ser uma subclasse de portfolio e se tornou um atributo opcional de portfolio. 

Essa alteração influenciou também na estrutura de market_data e market_data_mad, onde market_data_mad deixou de existir de maneira independente. Dessa forma ao construir o portfolio_mad, o MAD e o retorno esperados são calculados diretamente de market_data.

Assim ao chamar a função evaluate_mad() em portfolio, o objeto portfolio mad é construído se necessário e retorna o risco e o retorno esperado do portfólio.